### PR TITLE
[Criteria][Common] Implement SNUNet-Cd Loss

### PR DESCRIPTION
**Pull request by billylin609**

### CODE REVIEW

File: criteria/common/binary_dice_loss.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        y_pred: A tensor of shape [N, *]
+        y_true: A tensor of shape same with y_pred
+        reduction: Reduction method to apply, return mean over batch if 'mean',
+            return sum if 'sum', return a tensor of shape [N,] if 'none'
+    Returns:
+        Loss tensor according to arg reduction
+    Raise:
+        Exception if unexpected reduction
+    reference:
+    * https://github.com/hubutui/DiceLoss-PyTorch/blob/master/loss.py
```
Comment (danielmao2019) [inline criteria/common/binary_dice_loss.py:18]: Do they have a paper reference? I don't think it's a common approach to use the power `p`.

File: criteria/common/binary_dice_loss.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        self.p = p
+        self.reduction = reduction
+
+    def forward(self, y_pred: torch.Tensor, y_true: torch.Tensor) -> torch.Tensor:
+        assert y_pred.shape[0] == y_true.shape[0], "y_pred & y_true batch size don't match"
+        y_pred = y_pred.contiguous().view(y_pred.shape[0], -1)
+        y_true = y_true.contiguous().view(y_true.shape[0], -1)
+
+        num = torch.sum(torch.mul(y_pred, y_true), dim=1) + self.smooth
+        den = torch.sum(y_pred.pow(self.p) + y_true.pow(self.p), dim=1) + self.smooth
```
Comment (danielmao2019) [inline criteria/common/binary_dice_loss.py:32]: `torch.sum` applied to `+` or applied to each and then `+`?

File: criteria/common/__init__.py [hunk 3]
```diff
@@ -2,8 +2,12 @@
 CRITERIA.COMMON API
 """
 from criteria.common.focal_loss import FocalLoss
+from criteria.common.binary_dice_loss import BinaryDiceLoss
```
Comment (danielmao2019) [inline criteria/common/__init__.py:5]: Let's put this in `vision_2d`. Dice loss is most commonly used for semantic segmentation tasks, which is 2D vision.

**Review by danielmao2019 — CHANGES_REQUESTED** [review]

File: criteria/vision_2d/dice_loss.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        Args:
+            y_true: a tensor of shape [B, H, W].
+            y_pred: a tensor of shape [B, H, W]. Corresponds to
+                the raw output or y_pred of the model.
+        Returns:
+            dice_loss: the Sørensen–Dice loss.
+        """
+        assert isinstance(y_pred, torch.Tensor), "y_pred must be a PyTorch tensor."
+        assert isinstance(y_true, torch.Tensor), "y_true must be a PyTorch tensor."
+        assert y_pred.ndimension() == 3, "y_pred must have shape [B, H, W]."
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:29]: there will be a channels dimension.

File: criteria/vision_2d/dice_loss.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        assert y_true.ndimension() == 3, "y_true must have shape [B, H, W]."
+        assert y_pred.shape == y_true.shape
+
+        probas = torch.sigmoid(y_pred)
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:33]: in general, `y_pred` would be multi-class. So instead of `sigmoid`, use `softmax`. Moreover, the `softmax` should be done along `dim=1`.

File: criteria/vision_2d/dice_loss.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        # Ensure y_true is of the same type as y_pred
+        y_true = y_true.type_as(y_pred)
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:36]: it doesn't have to?

File: criteria/vision_2d/dice_loss.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+        # Compute intersection and union for Dice coefficient
+        intersection = torch.sum(probas * y_true, dim=(1, 2))
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:39]: summations are done along the spatial dimentions (height and width, which correspond to last two dims). but since `y_pred` should be of shape (N, C, H, W), the dim should be (-2, -1), or (2, 3).

File: criteria/vision_2d/dice_loss.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        cardinality = torch.sum(probas + y_true, dim=(1, 2))
+
+        # Compute Dice score
+        dice_score = (2. * intersection + self.eps) / (cardinality + self.eps)
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:43]: `dice_score` must be of shape (N,).

Comment (danielmao2019) [reply criteria/vision_2d/dice_loss.py:43]: add an assertion.

File: criteria/vision_2d/dice_loss.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        return 1 - dice_score.mean()
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:44]: Reduction would be `mean` in most cases. but better make this flexible by taking an argument `reduction` with default value `"mean"` in the `__init__`. For this, follow `torch.nn.CrossEntropyLoss`.

File: criteria/vision_2d/dice_loss.py (hunk 7 above)
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:39]: you need to do this per-class. So you'd have to make `y_true` one-hot encoded first.

File: criteria/vision_2d/snunet_criterion.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+from typing import Dict
+import torch
+import criteria
+from criteria.wrappers import SingleTaskCriterion
+from semantic_segmentation_criterion import SemanticSegmentationCriterion
+from dice_loss import DiceLoss
+
+
+class SNUNetCd(SingleTaskCriterion):
```
Comment (danielmao2019) [inline criteria/vision_2d/snunet_criterion.py:9]: Rename to `SNUNet_CD_Criterion`.

File: criteria/vision_2d/test_dice_loss.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    """Fixture to initialize the DiceLoss class."""
+    return DiceLoss(eps=1e-7)
+
+def test_dice_loss_multi_class(dice_loss):
+    """Test Dice loss calculation for multi-class segmentation."""
+    batch_size, height, width = 2, 64, 64
+    y_true = torch.randint(0, 2, (batch_size, height, width)).float()  # Binary mask
+    y_pred = torch.randn((batch_size, height, width))  # Raw logits
+
+    loss = dice_loss._compute_loss(y_pred, y_true)
```
Comment (danielmao2019) [inline criteria/vision_2d/test_dice_loss.py:16]: This is not testing the `DiceLoss` class's behavior, but rather testing a method of it. `DiceLoss` itself is callable.

File: criteria/vision_2d/dice_loss.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+from typing import Optional
+import torch
+
+class DiceLoss(torch.nn.Module):
```
Comment (danielmao2019) [inline criteria/vision_2d/dice_loss.py:4]: shouldn't be inheriting from `torch.nn.Module` but `SingleTaskCriterion`. Check how the other `vision_2d` criteria are implemented.

**Review by danielmao2019 — CHANGES_REQUESTED** [review]
